cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,489 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/tuple.h>
+#include <thrust/device_vector.h>
+#include <thrust/merge.h>
+
+#include <algorithm>
+#include <utility>
+#include <vector>
+#include <memory>
+#include <type_traits>
+
+#include <cudf/cudf.h>
+#include <cudf/types.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <rmm/thrust_rmm_allocator.h>
+#include <utilities/legacy/cuda_utils.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <cudf/null_mask.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+//#include <cudf/strings/detail/merge.cuh> // <- TODO: separate PR for strings support
+
+#include <cudf/detail/merge.hpp>
+
+namespace { // anonym.
```

Comment:
You mean encapsulate anonymous namespace inside `namespace cudf`? Hmm, not sure about that. I mean, yes, it can be done, but it's still a form of namespace pollution. 

I could use `using namespace cudf;` _inside_ each of the function blocks from the anonymous namespace, instead. Would that be acceptable?

Reply:
...or, even better: `using namespace cudf;` inside anonymous namespace's block header...

Reply:
@trevorsm7 , please confirm which option above you'd prefer. Thanks.

Reply:
I think encapsulating anonymous namespace inside namespace cudf is fine. I've seen it done this way in several other places in cudf. For example: https://github.com/rapidsai/cudf/blob/6f81a5ccd21abb245fba869045cd63a4cd6250cc/cpp/src/copying/copy.cpp#L28-L43

Reply:
Pending your comment, I've chosen to inject the `cudf` namespace into the anonymous one, since the latter uses it pretty much everywhere. This should not create too much of an issue since namespace injection is contained to a (very small) anonymous namespace block. Let me know if we can go ahead with this approach. Thanks.

Reply:
Ok, that's fine